cmd/dagger/query.go

```diff
@@ -0,0 +1,126 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"dagger.io/dagger"
+	"github.com/spf13/cobra"
+	"golang.org/x/term"
+)
+
+var queryCmd = &cobra.Command{
+	Use:                   "query [flags] [operation]",
+	Aliases:               []string{"q"},
+	DisableFlagsInUseLine: true,
+	Long:                  "Send API queries to a dagger engine\n\nWhen no document file, read query from standard input.",
+	Short:                 "Send API queries to a dagger engine",
+	Example: `
+dagger query <<EOF
+{
+  container {
+    from(address:"hello-world") {
+      exec(args:["/hello"]) {
+        stdout {
+          contents
+        }
+      }
+    }
+  }
+}
+EOF
+`,
+	Run:  Query,
+	Args: cobra.MaximumNArgs(1), // operation can be specified
+}
+
+func Query(cmd *cobra.Command, args []string) {
+	ctx := context.Background()
+	var operation string
+	if len(args) > 0 {
+		operation = args[0]
+	}
+
+	vars := getKVInput(queryVarsInput)
+
+	// Use the provided query file if specified
+	// Otherwise, if stdin is a pipe or other non-tty thing, read from it.
+	// Finally, default to the operations returned by the loadExtension query
+	var operations string
+	if queryFile != "" {
+		inBytes, err := os.ReadFile(queryFile)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+			os.Exit(1)
+		}
+		operations = string(inBytes)
+	} else if !term.IsTerminal(int(os.Stdin.Fd())) {
+		inBytes, err := io.ReadAll(os.Stdin)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+			os.Exit(1)
+		}
+		operations = string(inBytes)
+	}
+
+	result, err := doQuery(ctx, operations, operation, vars)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+
+	fmt.Printf("%s\n", result)
+}
+
+func doQuery(ctx context.Context, query, op string, vars map[string]string) ([]byte, error) {
+	opts := []dagger.ClientOpt{
+		dagger.WithWorkdir(workdir),
+		dagger.WithConfigPath(configPath),
+	}
+
+	if queryDebug {
+		opts = append(opts, dagger.WithLogOutput(os.Stderr))
+	}
+	c, err := dagger.Connect(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	defer c.Close()
+
+	res := make(map[string]interface{})
+	resp := &dagger.Response{Data: &res}
+	err = c.Do(ctx,
+		&dagger.Request{
+			Query:     query,
+			Variables: vars,
+			OpName:    op,
+		},
+		resp,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.Errors) > 0 {
+		return nil, resp.Errors
+	}
+
+	result, err := json.MarshalIndent(res, "", "    ")
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func getKVInput(kvs []string) map[string]string {
```

Comment:
Just FYI this ability to provide vars as a command line flag has a bunch of problems, biggest one being that you can't really provide a list: https://github.com/dagger/dagger/issues/3082

We were okay with it at the time since everything was in prototype phase, but I think we should either fix this or remove it for now before the release.

Reply:
`+1`

Reply:
working on this RN